### PR TITLE
[feature] adds serial ports for iso and clone builder

### LINKED ISSUE #41: Allow adding serial sockets to template
#### Description

We can add "serial0" as the vga display type, but as far as I can tell there isn't any way to create serial0 in packer.

#### Use Case(s)

This would allow configurations of serial ports for the template. Proxmox allows port 0-3 it seems.

#### Potential configuration

```
serial_port [0, 1, 3]
```

#### Potential References
Proxmox has a page about Serial Terminal here: https://pve.proxmox.com/wiki/Serial_Terminal

### PR DESCRIPTION
**Pull request by mabeett**
Adds `serials` option which allows user adding from 1 up to 4 serials ports to the packer VM.

Closes #41

### CODE REVIEW

**Comment by mabeett** [comment]
> Just an FYI @mabeett  - "elements" is misspelt.

Thanks for your feedback!

I've just corrected the typo. If you consider there could be a better writing for the doc just let me know. English is not my mother's tongue.

File: builder/proxmox/common/config.go
Comment (lbajolet-hashicorp) [inline builder/proxmox/common/config.go:60]: Given the limitations described in the documentation, we can probably add a few checks in the `Prepare` step for the plugin, so we ensure that only valid values are accepted before we attempt to create the devices.

Like ensuring there are 4 or less elements, and that their values are the expected shape (assuming this is relevant?)

Also I wonder, why the limit of 4? Is it something enforced by the API?

**Review by lbajolet-hashicorp — CHANGES_REQUESTED** [review]
The code looks OK to me overall, once the checks are implemented (or not, if they're not relevant), we can consider merging this.

Thanks @mabeett!

File: builder/proxmox/common/config.go
Comment (mabeett) [reply builder/proxmox/common/config.go:60]: Yes, The limit of 4 is imposed by the API. 
In case of using more than 4  the build fails with `Error creating VM: 400 Parameter verification failed` message.
During the next days I will work on your suggestion during my available spare time.

File: builder/proxmox/common/config.go
Comment (lbajolet-hashicorp) [inline builder/proxmox/common/config.go:237]: If you have no use for the index in any place, you can probably change this loop to
```suggestion
	for _, serial := range c.Serials {
```

This way `serial` will be each entry in the `c.Serials` array, and you don't have to access it through the index in the following block.

File: builder/proxmox/common/config.go
Comment (lbajolet-hashicorp) [inline builder/proxmox/common/config.go:234]: Just a nit to rephrase the error here, so that users have a better idea what's happening with their serials.

```suggestion
	    errs = packersdk.MultiErrorAppend(errs, fmt.Errorf("too many serials: %d serials defined, but proxmox accepts 4 elements maximum", len(c.Serials)))
```

**Review by lbajolet-hashicorp — APPROVED** [review]
Besides the last few nits I left, the code looks good to me, once addressed we can merge.

Pre-approving for later

File: builder/proxmox/common/config.go
Comment (mabeett) [reply builder/proxmox/common/config.go:61]: I've just pushed 2 commits and rebased the branch.
From my side I tested the changes manually the validation.

Comment (lbajolet-hashicorp) [reply builder/proxmox/common/config.go:61]: Thanks @mabeett, I left a review at the same time, just a few nits to address, regarding testing I think if we could validate at least the `Prepare` stuff automatically that'd be best.
Maybe consider adding this to `builder/proxmox/common/config_test.go`? Other than that, we can merge this, if you validated manually that's a good start.

**Comment by lbajolet-hashicorp** [comment]
Hi @mabeett,

I took the liberty of adding one commit on your branch to test that the configs work as advertised, locally it does pass, we'll just wait until CI confirms it, and we can merge this PR.

Thanks again for the contribution!

**Comment by mabeett** [comment]
Hi @lbajolet-hashicorp !, 

> I took the liberty of adding one commit on your branch to test that the configs work as advertised, locally it does pass, we'll just wait until CI confirms it, and we can merge this PR.

Thank you so much, 
my interventions on this repository are 100% best effort mostly 0% Clang knowledge. 
I'll take the commit as reference in case of future proposals. 

kind regards,

**Pull request merged**
